src/main/java/org/hobbit/core/components/AbstractTaskGenerator.java

```diff
@@ -155,7 +160,9 @@ public void run() throws Exception {
     @Override
     public void receiveGeneratedData(byte[] data) {
         try {
+            currentlyProcessedMessages.acquire();
             generateTask(data);
+            currentlyProcessedMessages.release(1);
```

Comment:
Tthe release here is not safe. Please add an additional `try-finally` block to ensure that the permit is released even if `generateTask(data)` throws an exception.